srml/system/src/lib.rs

```diff
@@ -755,19 +766,54 @@ impl<T: Trait> Module<T> {
 	}
 }
 
-/// Weight limit check and increment.
+/// resource limit check.
 #[derive(Encode, Decode, Clone, Eq, PartialEq)]
 pub struct CheckWeight<T: Trait + Send + Sync>(PhantomData<T>);
 
 impl<T: Trait + Send + Sync> CheckWeight<T> {
-	fn internal_check_weight(weight: Weight) -> Result<(), DispatchError> {
+	/// Checks if the current extrinsic can fit into the block with respect to block weight limits.
+	///
+	/// Upon successes, it returns the new block weight as a `Result`.
+	fn check_weight(info: DispatchInfo) -> Result<Weight, DispatchError> {
 		let current_weight = Module::<T>::all_extrinsics_weight();
-		let next_weight = current_weight.saturating_add(weight);
-		if next_weight > T::MaximumBlockWeight::get() {
-			return Err(DispatchError::Payment)
+		let maximum_weight = T::MaximumBlockWeight::get();
+		let limit = match info.class {
+			DispatchClass::Operational => maximum_weight,
+			DispatchClass::User => maximum_weight / 4
+		};
+		let added_weight = info.weight.min(limit);
+		let next_weight = current_weight.saturating_add(added_weight);
+		if next_weight > limit {
+			return Err(DispatchError::BadState)
 		}
-		AllExtrinsicsWeight::put(next_weight);
-		Ok(())
+		Ok(next_weight)
+	}
+
+	/// Checks if the current extrinsic can fit into the block with respect to block length limits.
+	///
+	/// Upon successes, it returns the new block length as a `Result`.
+	fn check_block_length(_info: DispatchInfo, len: usize) -> Result<u32, DispatchError> {
+		let current_len = Module::<T>::all_extrinsics_len();
+		let added_len = len as u32;
+		let next_len = current_len.saturating_add(added_len);
+		if next_len > T::MaximumBlockSize::get() {
```

Comment:
probably want to do the same as with weight and have 3/4 reserved for operational transactions.

Reply:
Sure 👍